File: tree/java/leetcode103_Binary_Tree_Zigzag_Level_Order_Traversal.java
//Given a binary tree, return the zigzag level order traversal of its nodes' values. (ie, from left to right, then right to left for the next level and alternate between).
//
//For example:
//Given binary tree [3,9,20,null,null,15,7],
//    3
//   / \
//  9  20
//    /  \
//   15   7
//return its zigzag level order traversal as:
//[
//  [3],
//  [20,9],
//  [15,7]
//]

package wenxinjie1;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Stack;

public class leetcode103_Binary_Tree_Zigzag_Level_Order_Traversal {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        Stack<TreeNode> stack = new Stack<TreeNode>();
        if (root == null) return res;
        stack.add(root);
        int flag = 1;
        while (!stack.empty()){
            Stack<TreeNode> temp = new Stack<TreeNode>();
            List<Integer> tem = new ArrayList<Integer>();
            for (TreeNode node : stack){
                tem.add(node.val);
                if (node.left != null) temp.add(node.left);
                if (node.right != null) temp.add(node.right);
            }
            if (flag == 1) res.add(tem);
            else {
                Collections.reverse(tem);
                res.add(tem);
            }
            flag = -flag;
            stack = temp;
        }
        return res;
        
    }
}

//Time: O(n)
//Space: O(n)
//Difficulty: medium

File: tree/python/leetcode103_Binary_Tree_Zigzag_Level_Order_Traversal.py
# Given a binary tree, return the zigzag level order traversal of its nodes' values. (ie, from left to right, then right to left for the next level and alternate between).

# For example:
# Given binary tree [3,9,20,null,null,15,7],
#     3
#    / \
#   9  20
#     /  \
#    15   7
# return its zigzag level order traversal as:
# [
#   [3],
#   [20,9],
#   [15,7]
# ]

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root: return []
        res, cur, flag, stack = [], [], 1, [root]
        while stack :
            for i in range(len(stack)):
                tem = stack.pop(0)
                if tem.left != None: stack.append(tem.left)
                if tem.right != None: stack.append(tem.right)
                cur.append(tem.val)
            res.append(cur[::flag])
            flag *= -1
            cur = []
        return res

# Time: O(n)
# Space: O(n)
# Difficulty: medium